### PR TITLE
Add relations between entities

### LINKED ISSUE #10: Add Connections between Entities
Requires Entities issue to be done

Add relationships between entities as decided
Implement add/remove methods on entities

### PR DESCRIPTION
**Pull request by helenahalldiniths**
This will fix #10.
The following connections have been added:

1. Crime -> Category (unidirectional, ManyToOne)
2. Crime -> Address (unidirectional, ManyToOne)
3. Criminal -> Address (unidirectional, ManyToOne)
4. Victim -> Address (unidirectional, ManyToOne)
5. Crime <-> Victim (bidirectional, ManyToMany, Crime is parent)
6. Crime <-> Criminal (bidirectional, ManyToMany, Crime is parent)

### CODE REVIEW

File: src/main/java/se/iths/crimedatabase/entity/Crime.java [hunk 1]
```diff
@@ -3,22 +3,46 @@
...
     private Long id;
+
     private String name;
+
     @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm")
     @JsonFormat(pattern = "yyyy-MM-dd HH:mm")
     private LocalDateTime time;
 
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+    @JoinColumn
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/entity/Crime.java]: Do we need @JoinColumn?

File: src/main/java/se/iths/crimedatabase/entity/Crime.java [hunk 2]
```diff
@@ -3,22 +3,46 @@
...
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+
     private String name;
+
     @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm")
     @JsonFormat(pattern = "yyyy-MM-dd HH:mm")
     private LocalDateTime time;
 
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/entity/Crime.java]: If we remove a crime we also remove a category? (cascadetype)

File: src/main/java/se/iths/crimedatabase/entity/Crime.java [hunk 3]
```diff
@@ -3,22 +3,46 @@
...
+    @JoinColumn
+    private Category category;
+
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/entity/Crime.java]: Same here, remove crime and address together?

File: src/main/java/se/iths/crimedatabase/entity/Crime.java [hunk 4]
```diff
@@ -3,22 +3,46 @@
...
+    private Address address;
+
+    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REFRESH})
+    private Set<Victim> victims = new HashSet<>();
+
+    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REFRESH})
+    private Set<Criminal> criminals = new HashSet<>();
+
+    public void addVictim(Victim victim) {
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/entity/Crime.java:35]: Do we need remove methods?

File: src/main/java/se/iths/crimedatabase/entity/Crime.java [hunk 5]
```diff
@@ -3,22 +3,46 @@
...
     private LocalDateTime time;
 
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+    @JoinColumn
+    private Category category;
+
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+    private Address address;
+
+    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REFRESH})
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/entity/Crime.java]: I believe cascade should be set on entity that is not mapping the relation. In this case Victim and Criminal entity fields?

**Review by LordRekishi — CHANGES_REQUESTED** [review]

File: src/main/java/se/iths/crimedatabase/entity/Crime.java (hunk 1 above)
Comment (helenahalldiniths) [reply src/main/java/se/iths/crimedatabase/entity/Crime.java]: No we don't! It is default when using ManyToOne! My bad :)

File: src/main/java/se/iths/crimedatabase/entity/Crime.java (hunk 4 above)
Comment (helenahalldiniths) [reply src/main/java/se/iths/crimedatabase/entity/Crime.java:35]: Remove should never be used when using ManyToMany!  

You can read about it here:
https://thorben-janssen.com/avoid-cascadetype-delete-many-assocations/

Comment (helenahalldiniths) [reply src/main/java/se/iths/crimedatabase/entity/Crime.java:35]: Sorry, you refered to a method like "removeVictim" not cascade type. 
I think it depends on what we want to be able to do with our application later on. For now I think add is enought?

File: src/main/java/se/iths/crimedatabase/entity/Crime.java (hunk 5 above)
Comment (helenahalldiniths) [reply src/main/java/se/iths/crimedatabase/entity/Crime.java]: That sounds right, I will move it :)

File: src/main/java/se/iths/crimedatabase/entity/Crime.java (hunk 2 above)
Comment (helenahalldiniths) [reply src/main/java/se/iths/crimedatabase/entity/Crime.java]: I will do some diging and get back to you!

Comment (helenahalldiniths) [reply src/main/java/se/iths/crimedatabase/entity/Crime.java]: I didn't find any good resources... but you are correct in that we don't want to remove a category if we remove a crime that has that category. However, my thought was that the cascade typ "remove" (when used on a manyToOne relationships) only removes it if it doesn't have any connections left, i.e if we dont have any crimes with the category "Theft" than "Theft" gets removed from the category tabel. 

I'm not sure how it work, I think I will ask Martin if I don't find any good info.

Comment (LordRekishi) [reply src/main/java/se/iths/crimedatabase/entity/Crime.java]: Please check and we will see

File: src/main/java/se/iths/crimedatabase/entity/Crime.java (hunk 4 above)
Comment (LordRekishi) [reply src/main/java/se/iths/crimedatabase/entity/Crime.java:35]: It's enough for now yes. So we can leave this like it is

File: src/main/java/se/iths/crimedatabase/entity/Crime.java (hunk 2 above)
Comment (helenahalldiniths) [reply src/main/java/se/iths/crimedatabase/entity/Crime.java]: Martin said "good question" and than confessed that he doesn't know... 

After disscussing it a little we desided that it is best to remove the cascade type "Remove" and let there be categories that doesn't have any crimes connected to them if all crimes that has a category are removed.

**Review by LordRekishi — APPROVED** [review]

**Pull request merged**